templates/terraform/cloudsql/main.tf

```diff
@@ -0,0 +1,82 @@
+# NOTE: This is needed because CloudSQL doesn't allow DB names to be reused
+# within the week that they're deleted, so whenever we want to tear down and
+# recreate the instance we need to generate a new name (vs waiting a week).
+resource random_id cloudsql_random_id {
+  byte_length = 8
+}
+
+# Create a single DB instance to back our applications.
+# NOTE: We might need multiple instances in prod, if it turns out there's too
+# much resource contention sharing a single instance.
+resource google_sql_database_instance postgres {
+  provider = google.target
+  depends_on = [var.dependencies]
+
+  name = "${var.name_prefix}-postgres-${random_id.cloudsql_random_id.hex}"
+  database_version = var.postgres_version
+
+  settings {
+    activation_policy = "ALWAYS"
+    pricing_plan = "PER_USE"
+    replication_type = "SYNCHRONOUS"
+    tier = "db-custom-${var.cpu}-${var.ram}"
+    user_labels = var.labels
+
+    backup_configuration {
+      binary_log_enabled = false
+      enabled = true
+      start_time = "06:00"
+    }
+
+    ip_configuration {
+      ipv4_enabled = true
+      require_ssl = true
+      authorized_networks {
+        name = "Broad"
+        value = "69.173.64.0/18"
+      }
+    }
+  }
+}
+
+resource random_id db-password {
+  for_each = toset(var.user_names)
+
+  byte_length = 16
+}
+
+resource google_sql_user db-user {
+  for_each = toset(var.user_names)
+
+  provider = google.target
+  name = "${var.name_prefix}-${each.value}"
+  password = random_id.db-password[each.value].hex
+  instance = google_sql_database_instance.postgres.name
+}
```

Comment:
These user/password pairs should also be written to Vault. I think one secret per user makes sense, with:
1. The "leaf" of the secret being the username
2. One key in the secret for the password

Reply:
by "leaf" do you mean the last part of the path? So like "${var.vault_prefix}/cloudsql/users/<put_username_here>"

Reply:
Yes